File: backend/blocky/api/listeners.py
"""`/api/listeners`：把畫布上的 hat 接上／斷開（§9、P1 第 4 步第 3 段）。

**是 `/api/listeners` 而不是 `/api/triggers`**：這條路管的是「這個專案現在有
沒有在聽」，而 §9 的 trigger 是被聽的那一端（cron、webhook、積木包的 hat）。
P2 的 Trigger Manager 進來時，它管的東西會住在 `/api/triggers`，兩者不該撞名。

**以專案為單位，不是以 hat 為單位。** 一個專案的 hat 集合來自它的 IR，而
「使用者要聽哪幾顆」不是一個獨立的決定——他要的是「這份流程有沒有在跑」。
§9.2 的「diff 新舊 hat 集合、只重啟有變動的」是同一個粒度往下切，屬於 P2。
"""

from __future__ import annotations

from typing import Any

from fastapi import APIRouter, Body, HTTPException, Request
from pydantic import BaseModel

from blocky.api.errors import invalid_ir
from blocky.errors import BlockyError, ValidationError
from blocky.runs import ProjectNotFound
from blocky.runs.listeners import ListenerManager

router = APIRouter(prefix="/api/listeners", tags=["listeners"])


class ListenRequest(BaseModel):
    projectId: str


def _listeners(request: Request) -> ListenerManager:
    return request.app.state.listeners


@router.get("")
async def list_listeners(request: Request) -> list[dict[str, Any]]:
    return [listener.summary() for listener in _listeners(request).list()]


@router.post("", status_code=201)
async def start_listening(request: Request, body: ListenRequest = Body(...)) -> dict[str, Any]:  # noqa: B008
    """接上這個專案的 hat。**已經在聽就原樣回去**，不是 409。

    前端的「執行」會順手打這一條（使用者不必知道那是兩件事），所以「已經開著」
    是最常見的情況，不是例外狀況。
    """
    try:
        listener = await _listeners(request).start(body.projectId)
    except ProjectNotFound:
        raise HTTPException(
            status_code=404,
            detail={"message": f"找不到專案 {body.projectId}", "hint": "先存檔再監聽"},
        ) from None
    except ValidationError as e:
        raise invalid_ir(e) from None
    except BlockyError as e:
        # 積木包接不上（token 不對、包壞了）。422 而不是 500：錯的是這份專案
        # 用到的東西，不是後端。
        raise HTTPException(status_code=422, detail=e.to_dict()) from None
    return listener.summary()


@router.delete("/{project_id}", status_code=204)
async def stop_listening(project_id: str, request: Request) -> None:
    """斷開。**沒在聽也是 204**：使用者要的結果是「現在沒在聽」，而那已經成立。"""
    await _listeners(request).stop(project_id)

File: packages/editor/src/api/listeners.ts
/**
 * `/api/listeners`：把畫布上的 hat 接上／斷開（§9、P1 第 4 步第 3 段）。
 *
 * **跟「執行」是兩件事，所以是兩條路。** 執行是「現在跑一次」，監聽是「一直
 * 聽著，外面發生事情就跑」——後者沒有結束時間，也不屬於任何一個 runId。共用
 * `/api/runs` 的話，「停止」到底是停哪一個就講不清楚了。
 *
 * 後端不是 §9 的 Trigger Manager，是它的前身：監聽活在後端的記憶體裡，重啟就
 * 沒了（`backend/blocky/runs/listeners.py` 開頭那張表列了差在哪）。
 */
import { toApiError } from './client';

export interface ListenerSummary {
  projectId: string;
  startedAt: string;
  /** 接上的 hat opcode。空陣列 = 這份畫布上沒有 hat，不是失敗。 */
  hats: string[];
  /** 監聽期間沒有 Run 可以歸屬的錯誤（連線斷了、token 不對）。 */
  errors: string[];
}

export async function startListening(projectId: string): Promise<ListenerSummary> {
  const res = await fetch('/api/listeners', {
    method: 'POST',
    headers: { 'Content-Type': 'application/json' },
    body: JSON.stringify({ projectId }),
  });
  if (!res.ok) throw await toApiError(res, `POST /api/listeners → ${res.status}`);
  return (await res.json()) as ListenerSummary;
}

/** 斷開。沒在聽也是 204——要的結果是「現在沒在聽」，而那已經成立。 */
export async function stopListening(projectId: string): Promise<void> {
  const res = await fetch(`/api/listeners/${encodeURIComponent(projectId)}`, { method: 'DELETE' });
  if (!res.ok) throw await toApiError(res, `DELETE /api/listeners/${projectId}`);
}

export async function listListeners(): Promise<ListenerSummary[]> {
  const res = await fetch('/api/listeners');
  if (!res.ok) throw await toApiError(res, `GET /api/listeners → ${res.status}`);
  return (await res.json()) as ListenerSummary[];
}
